src/game.cpp

```diff
@@ -13210,11 +13235,11 @@ void game::shift_monsters( const int shiftx, const int shifty, const int shiftz
         if( shiftx != 0 || shifty != 0 ) {
             critter.shift( shiftx, shifty );
         }
-        // TODO: with z-levels, this can be removed, instead shift the critter
-        // along shiftz, too. Than make the 3D-inbounds check.
-        if( shiftz == 0 && m.inbounds( critter.posx(), critter.posy() ) ) {
+
+        if( shiftz != 0 && m.has_zlevels() && m.inbounds( critter.pos() ) ) {
```

Comment:
This logic is a bit flawed. It will despawn the monster whenever shiftz is 0.

However, funnily this does not affect the game at all because the monster is immediately re-spawned.

If you activate the debug messages and move across a submap boundary (which triggers the map shifting), you'll see the debug message from the monster upgrade stuff ("Last loaded: 0" and "Upgrade time less"). The monster-upgrade stuff is only triggered through `game::add_zombie` (so that function is definitely called).


Btw. and totally unrelated: the monster upgrade is done via a `monster::update_check` function, but all the variables of the `mtype` have the word upgrade in it. So what is it: an up**date** or a up**grade**? This should be a bit more consistent, it's very confusing.